### PR TITLE
Intrepid2: Add Hierarchical Basis (Hypercube Topologies)

### PR DESCRIPTION
**Pull request by CamelliaDPG**
@trilinos/intrepid2 

## Description
Added implementations of hierarchical bases on hypercube topologies, following work by Fuentes et al. (though for various reasons we do not include support for their "orientation-embedding").   See [Orientation Embedded High Order Shape Functions for the Exact Sequence Elements of All Shapes](https://arxiv.org/ct?url=https%3A%2F%2Fdx.doi.org%2F10.1016%2Fj.camwa.2015.04.027&v=80f505c5).

Motivated by the mathematical construction of Fuentes et al, the implementation of the 2D and 3D bases are templated on the 1D basis definitions.  This allows a reduced overall code footprint, and also will allow us to reimplement several of the existing nodal bases with less code, more reuse, and explicitly defined tensor-product structures.  Going forward, we hope to exploit tensor-product structures to improve performance.

In addition to implementing the basis functions defined by Fuentes et al., we include variants that relax the usual requirement that H(grad) includes "vertex" functions on the line, etc.  These "DG" bases are identified with the interior of the cell.  This affords us the flexibility to maintain a strict polynomial hierarchy, starting with the constant basis.  That in turn allows simpler enforcement of conservation, which is one of the motivations for this implementation.

We also introduce here the notion of a BasisFamily.  This acts as a factory for a set of related basis functions.  To start, we support the hypercube topologies; later, we will add support for "all types" (triangles, pyramids, wedges, tetrahedra).  We will also be adding hierarchical bases for those topologies, again following Fuentes et al. for the construction.

## How Has This Been Tested?
We have run extensive cross-code verification with the ESEAS Fortran library (a companion to the paper cited above) to confirm that our implementation agrees with theirs.  Eventually, we may include these verification tests alongside the tests included in this PR, but this would require the ESEAS library to be included with Trilinos, which in turn requires an appropriate (BSD) open source license.  We have contacted the authors about making that licensing change.

Additionally, there are a host of unit tests that exercise various features of this implementation, and compare basis values against analytically expected values.

## Checklist
- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_IntegratedLegendreBasis_HGRAD_LINE.hpp [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Kyungjoo Kim  (kyukim@sandia.gov),
+//                    Mauro Perego  (mperego@sandia.gov), or
+//                    Nate Roberts  (nvrober@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+/** \file   Intrepid2_IntegratedLegendreBasis_HGRAD_LINE.hpp
```
Comment (mperego) [inline packages/intrepid2/src/Discretization/Basis/Intrepid2_IntegratedLegendreBasis_HGRAD_LINE.hpp]: I would reference the Fuentes paper here as well, and maybe briefly describe how the HGRAD line polynomials are built, that they are integrated polynomials, and the difference between the CG and DG case

File: packages/intrepid2/unit-test/Discretization/Basis/HierarchicalBases/AnalyticPolynomialsMatchTests.cpp [hunk 2]
```diff
@@ -0,0 +1,689 @@
...
+#include "Kokkos_Core.hpp"
+
+namespace
+{
+  using namespace Intrepid2;
+  
+  template<typename ExecutionSpace=Kokkos::DefaultExecutionSpace,
+           typename OutputScalar = double,
+           typename PointScalar  = double>
+  void testHierarchicalHGRAD_LINE_MatchesAnalyticValues(Intrepid2::EOperator op, const double tol, Teuchos::FancyOStream &out, bool &success)
```
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/unit-test/Discretization/Basis/HierarchicalBases/AnalyticPolynomialsMatchTests.cpp]: It had not; I've just pushed some changes, such that this is now tested with (DFad,DFad) as well.  This does now pass.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HCURL_QUAD.hpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+// Questions? Contact Kyungjoo Kim  (kyukim@sandia.gov),
+//                    Mauro Perego  (mperego@sandia.gov), or
+//                    Nate Roberts  (nvrober@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+/** \file   Intrepid2_DerivedBasis_HCURL_QUAD.hpp
+    \brief  Implementation of H(curl) basis on the quadrilateral that is templated on H(vol) and H(grad) on the line.
+    \author Created by N.V. Roberts.
```
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HCURL_QUAD.hpp]: I've now added such documentation to this and the other `DerivedBasis` classes.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HDIV_HEX.hpp [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+// Questions? Contact Kyungjoo Kim  (kyukim@sandia.gov),
+//                    Mauro Perego  (mperego@sandia.gov), or
+//                    Nate Roberts  (nvrober@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+/** \file   Intrepid2_DerivedBasis_HDIV_HEX.hpp
+    \brief  Implementation of H(div) basis on the hexahedron that is templated on H(vol) and H(grad) on the line.
+    \author Created by N.V. Roberts.
```
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HDIV_HEX.hpp]: Done.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HDIV_QUAD.hpp [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+//
+// Questions? Contact Kyungjoo Kim  (kyukim@sandia.gov),
+//                    Mauro Perego  (mperego@sandia.gov), or
+//                    Nate Roberts  (nvrober@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+/** \file   Intrepid2_DerivedBasis_HIV_QUAD.hpp
+    \brief  Implementation of H(div) basis on the quadrilateral that is templated on H(vol) and H(grad) on the line.
```
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HDIV_QUAD.hpp]: Done.

**Comment by mperego** [comment]
@CamelliaDPG Let me know when you are done and you'd like me to approve. Thanks.

**Comment by jmgate** [comment]
@CamelliaDPG, are you intending to modify 3150 files in this PR?

**Comment by CamelliaDPG** [comment]
> @CamelliaDPG, are you intending to modify 3150 files in this PR?

@jmgate, no; I just made an error in squashing in my fork.  It's going to take a while to resolve, I think…

**Comment by CamelliaDPG** [comment]
> @CamelliaDPG Let me know when you are done and you'd like me to approve. Thanks.

@mperego, this should now be ready to go.  I've squashed things down to a single commit, and I believe I've addressed all your comments.

**Pull request merged**